posts/2023-04-13-rust-by-examples.adoc

```diff
@@ -0,0 +1,726 @@
+
+We could just return the found &str, and save the creation of the String, but the following code will fail to compile:
+[source,rust]
+----
+fn return_match_borrow(pattern: &str, lines: &Vec<&str>) -> Option<&str> {
+    lines
+        .iter()
+        .find(|&line| line.contains(&pattern))
+        .map(|&line| line)
+}
+// error: missing lifetime specifier
+// this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from `pattern` or one of `lines`'s 2 lifetimes
+----
+
+We'll need to define the lifetime of the response and bind it to the lifetime of a function argument. Lifetimes are defined as labels in the format of ``'x``, where x identifies the lifetime:
+[source,rust]
+----
+fn return_match_borrow<'a>(pattern: &str, lines: &'a Vec<&str>) -> Option<&'a str> {
+....
+----
+
+We have a special case, our input are static values and we can also use the special lifetime label ``'static``:
+[source,rust]
+----
+fn return_match_borrow(pattern: &str, lines: &'static Vec<&str>) -> Option<&'static str> {
+....
+----
+
+To me, ownership and lifetimes are the most cumbersome part of Rust development, but this is the price to pay to avoid garbage collection.
+``Object`` in Java or ``AnyRef`` Scala are passed on as references and we do not need to worry about who owns them. Execution is as efficient as it can be, and Rust does not have much performance advantage over Java.
+This is possible, because the JVM counts the references to each object and the memory allocated to them is freed when there is no more reference remaining. However, garbage collection is expensive, it consumes memory and processor cycles. It is done periodically, in multiple levels, making execution times fluctuate.
+
+=== Error handling
+
+There is no ``Exception`` in Rust. Non-fatal errors are usually managed by setting a ``Result<R, E>`` response type for the functions. On success, the function returns ``Ok<R>`` on failure returns ``Err<E>``. And yes, it is the opposite order to ``Either<E, R>`` of Scala. There are constructs in Rust to make Result handling convenient, you can read the https://doc.rust-lang.org/book/ch09-00-error-handling.html[guide^] for more info.
+Fatal errors, ignorant error handling practices or calling the ``panic!`` macro will cause a 'panic'. By default, these panics will print a failure message, unwind, clean up the stack, and quit. There is no way to recover after a panic.
+
+== Functional Programming Features
+
+We have seen iterators, closures and Rust's love for immutable variables in the predious sections. Rust gives us the choice to write procedural or FP style code. They are both fine and there is negligeable performance difference between good for-loop and iterator based solutions. FP code is often easier to understand and naturally efficient, but loops can give greater control.
```

Comment:
* You mention _We have seen iterators, closures_ but there are in the next section
* Typo: `predious`

Reply:
I've simplified this paragraph